Draft optics implementation

AriFordsham opened the pull request:
Initial draft implementation.

AriFordsham committed:
Typo

AriFordsham committed:
Initial lenses

AriFordsham commented:
What obstacles are there to 'efficiently?'

I'm planning on implementing profunctor optics in a way that makes sense for Plutarch lenses.

- As much work should be done in Haskell, at Plutarch compile time.
- I need to ensure not to introduce superfluous `pcon`/`pmatch` into the body of combinators and lenses themselves.
- Is there a risk of redundant `pcon`/`pmatch` in a composition of optics, which the composition function needs to deal with? It doesn't seem obvious to me that there is.
- The general problem of redundant matches might be unsolvable in library, and may need an optimizer pass.

Are there any efficiency considerations I am missing?

L-as commented:
Consider combining `Lens' a b` and `Lens' a c` into `Lens' a (b, c)`. How do you this efficiently?
Relying on the backend to optimise it is not good design.

AriFordsham committed:
Initial optics

AriFordsham commented:
- Optics for `EPair` and `EEither`
- type aliases `PLens` and `PPrism`
- typeclass `PChoice` for `PPrism`
- some useful combintors in `Core`

AriFordsham committed:
PChoice

AriFordsham committed:
Pair and Either optics lenses with new Profunctor and Lens hierarchy

AriFordsham committed:
Runner for Optional

AriFordsham committed:
Traversal attempt

AriFordsham committed:
Fixed instances

AriFordsham committed:
Explicit CPSs - typing problem

AriFordsham committed:
continuation-based optics

AriFordsham commented:
We have optics!

- Plutarch-free profunctor and optics stacks based on the `Cont` monad - essentially replacing every `a -> b` with `a -> Cont r b`
- Optics for `EPair` and `EEither`. These essentially use Haskell `Either` in Plutarch!

AriFordsham committed:
Fix

AriFordsham committed:
Cleanup

L-as commented:
Why change the warnings?
Make sure to use GHC 9.2.4. You can do `nix develop` to get the correct GHC.

AriFordsham commented:
Happy to change that back. I use `partial-type-signatures` warnings with HLS quite heavily in my workflow.
`shell.nix` on `master` has `ghc923`, which is what I'm using.

L-as commented on Plutarch/CPS/Profunctor.hs:
```diff
@@ -0,0 +1,68 @@
+newtype CStar r f a b = CStar { runCStar :: a -> Cont r (f b) }
+  
+class CProfunctor r p where
+  cdimap :: (a -> Cont r b) -> (c -> Cont r d) -> p b (Cont r c) -> p a (Cont r d)
```
Maybe this should still be `p b c` and `p a d`.

L-as replied:
Then `CProfunctor r ->` would not be implementable, but you could make a newtype where it is implementable.

L-as commented:
e.g.?

AriFordsham commented:
Not having `r` on the right of an arrow breaks a lot of inference and requires many type applications. In addition, something hard broke, I can't remember what. It seemed fixable, but not worth the trouble. I'm happy to try again if you think it's worth it.

L-as commented:
In general, type applications are a code smell, along with AllowAmbiguousTypes IMO.

AriFordsham committed:
Current progress

AriFordsham committed:
Merge branch 'master' into ari/optics

AriFordsham committed:
Traversals with List

AriFordsham committed:
Formatting

AriFordsham committed:
Merge branch 'master' into ari/optics

AriFordsham committed:
Updates

AriFordsham committed:
PList traversal

AriFordsham committed:
Prototype Haskell Fix Traversal combinator

Not actually a traversal, but allows easy definition of one.

With example of use

AriFordsham committed:
Fix transcribed to Plutarch

AriFordsham committed:
Merge branch 'master' into ari/optics

AriFordsham committed:
Current progress

AriFordsham committed:
Merge branch 'ari/simplified-generics' into ari/optics

AriFordsham committed:
Merge branch 'ari/simplified-generics' into ari/optics

AriFordsham committed:
Current progress

AriFordsham committed:
Merge branch 'ari/simplified-generics' into ari/optics

AriFordsham committed:
Current progress

AriFordsham committed:
Merge branch 'ari/simplified-generics' into ari/optics

AriFordsham committed:
Implemented `pconstrained`

AriFordsham commented:
`PConstrained` implemented

AriFordsham committed:
Completed `traverse`

AriFordsham commented:
- Completed `ptraverse` for types with one parameter
- Implemented `plist` in terms of `ptraverse`

We won't really know if this works until we have a backend to test with.

L-as commented:
Amazing! Perhaps we should work on getting a backend to work again.

L-as commented:
Plutarch has changed too much for this to be mergible.
